src/server/pkg/storage/kv/cache.go

```diff
@@ -0,0 +1,103 @@
+package kv
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+	log "github.com/sirupsen/logrus"
+)
+
+var _ Store = &WriteThroughCache{}
+
+type WriteThroughCache struct {
+	slow, fast Store
+
+	mu    sync.Mutex
+	cache *simplelru.LRU
+}
+
+func NewWriteThroughCache(slow, fast Store, size int) *WriteThroughCache {
+	c := &WriteThroughCache{
+		slow: slow,
+		fast: fast,
+	}
+	var err error
+	c.cache, err = simplelru.NewLRU(size, c.onEvict)
+	if err != nil {
+		panic(err)
+	}
+	return c
+}
+
+func (c *WriteThroughCache) GetF(ctx context.Context, p string, cb func([]byte) error) error {
+	c.mu.Lock()
+	if _, exists := c.cache.Get(p); exists {
+		c.mu.Unlock()
+		if err := c.fast.GetF(ctx, p, cb); err != nil {
```

Comment:
Is this a safe operation? It would require that the backing store either blocks a delete on a full read, or appropriately errors if the full read cannot be performed. If that is not the case, then we probably need a path lock similarly to the other implementation. If it is safe, then the other implementation probably does not need the path lock.

Reply:
I think you are correct. It does have the same problem as the other implementation.  But in this case "full read" is guaranteed to be as quick as possible because it's into memory.  Whereas a full read in the other implementation can take arbitrarily long, if there is a slow reader. Improper use (failure to close the reader) would cause the fast layer to grow unbounded.  In this implementation the only advantage is that the backing store doesn't have to track open readers.  Once the content is in a buffer you can forget about it.

I looked more into how files work on Linux: deleting a file does not seem to cause readers with pre-existing file descriptors to error.
https://stackoverflow.com/questions/2028874/what-happens-to-an-open-file-handle-on-linux-if-the-pointed-file-gets-moved-or-d
So that should work for our use case.  I'm not sure if all the object stores also work this way. It's difficult to find documentation about this specific case.